scipy/special/xsf/sphd_wave.h

```diff
@@ -62,14 +62,20 @@ void prolate_aswfa_nocv(T m, T n, T c, T x, T &s1f, T &s1d) {
     int_n = (int) n;
     eg = (T *) malloc(sizeof(T) * (n - m + 2));
     if (eg == NULL) {
-        set_error("prolate_aswfa_nocv", SF_ERROR_OTHER, "memory allocation error");
+        set_error("prolate_aswfa_nocv", SF_ERROR_NOMEM, "memory allocation error");
         s1d = std::numeric_limits<T>::quiet_NaN();
         s1f = std::numeric_limits<T>::quiet_NaN();
         return;
     }
     specfun::segv(int_m, int_n, c, kd, &cv, eg);
-    specfun::aswfa(x, int_m, int_n, c, kd, cv, &s1f, &s1d);
+    specfun::Status status = specfun::aswfa(x, int_m, int_n, c, kd, cv, &s1f, &s1d);
```

Comment:
Thanks for this PR @WarrenWeckesser and apologies for this quick-fire question. I might be saying something stupid since it's been a while I looked at these but here is what I was mumbling in the original issue; 

Would it not be better (and possibly higher performant) if we do the malloc calloc here and pass the pointer to aswfa as an extra argument instead of handling it within aswfa? Obviously it will require a change in the function body but then at least it would have the memory error happening here once as it is happening with `segv`? 

I don't know if I am writing it in the right spot of the code but what I am thinking is that the ufunc allocates the memory once and keeps sending it in to aswfa invocation across each element call

Reply:
That's what I was thinking too @ilayn. Ideally, I think the memory allocations should happen as high in the call stack they can, under the constraint that we need to know the size of the arrays to be allocated. I'd be willing to restructure the code to make that happen.

Reply:
Yes, that is definitely possible.  We could modify the functions in `specfun.h` to require preallocated work arrays.  This wouldn't be an immediate benefit, however, because *this* function (`prolate_aswfa_nocv()`) is still called for every set of scalar inputs, so the same repeated alloc and free is still happening.  To minimize memory allocations, a modification to the ufunc generation infrastructure is needed, so that the memory is allocated before entering the ufunc loop and freed after the loop.  To be specific, the loop that I'm talking about is one of

https://github.com/scipy/scipy/blob/bc0d3271b1243d9c0b3d05f3c2ac67fe4c5eb685/scipy/special/xsf/numpy.h#L819-L838

or

https://github.com/scipy/scipy/blob/bc0d3271b1243d9c0b3d05f3c2ac67fe4c5eb685/scipy/special/xsf/numpy.h#L785-L805

For this example, the `func` that is called in each iteration of the for-loop is `prolate_aswfa_nocv()`.  To minimize the repeated alloc/free, the allocation would need to happen before the for-loop is entered, and the free after the loop is exited.

(This wouldn't eliminate all repeated alloc/free, because when the inputs to the ufunc are higher-dimensional arrays, `loop()` itself is called multiple times by the NumPy ufunc looping machinery.)

This only works if the size of the required work array does not depend on the input values, which is not true for all the functions that do heap allocations.  When the size depends on the value, it might be necessary to do an initial pass over the data that was passed to `loop()` to find the worst case, and allocate that much before entering the main loop.  Or maybe something more clever is possible.

@izaid might have some ideas on how to cleanly incorporate some way of doing this in the C++ code.  Maybe something like pre- and post-loop functions that can be provided in the call that generates a ufunc?  It will need some way to ensure that the arrays that are generated in the pre-loop can be passed to the core function.  Or, perhaps a completely different approach would be that some ufuncs are generated from an explicitly implemented loop function (instead of just the scalar function); then we could do whatever is necessary before and after the loop.

Without such a modification to the ufunc generation code, there is no real benefit to moving the allocation up one level of function call.  Moving the allocation out of the function where it is needed  requires making implementation details of `aswfa()` (for example) available outside the function, which--in general--is something we would like to avoid.  But we definitely want to avoid the repeated alloc/free, so the change is worthwhile, if we anticipate that the required changes to the ufunc generation infrastructure will eventually be made.

Reply:
Great thoughts. This sort of thing has been on my mind too, being able to move the allocations up out of the loops, and it's also something Irwin has brought up before. If we can do it, I'm all for it.

For me, the benefit of moving the allocations up in the stack is small but meaningful. It's that it makes it easy to call `set_error` with the actual name of the function, instead of some internal thing like `prolate_awsfa`, without having to pass status codes up through the stack.

Reply:
Sorry, I'm not sure I understand your last comment.  Whether the allocation failure is detected in `aswfa()` or in `prolate_aswfa_nocv()` does not affect the name that is passed to `set_error()`.  We can pass whatever we like for the name.  As you can see here, `set_error()` is only called in `prolate_aswfa_nocv()`.  Currently I set the name in `set_error()` to `"prolate_aswfa_nocv"` because that is what the other existing calls to `set_error()` in this function use.  It would probably make sense to change all those to `"pro_ang1"`, since that is the name of the corresponding ufunc in the public Python API, as set here:

https://github.com/scipy/scipy/blob/bc0d3271b1243d9c0b3d05f3c2ac67fe4c5eb685/scipy/special/_special_ufuncs.cpp#L793-L796

Reply:
>  To minimize memory allocations, a modification to the ufunc generation infrastructure is needed, so that the memory is allocated before entering the ufunc loop and freed after the loop. 

Ah OK. Then it looks like a bigger surgery than I initially thought.

Reply:
Sorry @WarrenWeckesser. I haven't looked at this code in a while and didn't recall that this function was only called by one other function. I remember in amos, and possiblyy in specfun, there are helper functions which are called by multiple functions, sometimes several layers down in the stack, and I remember thinking it will be annoying to have to propagate the function names to the top if something goes wrong.

In this case, where the function is only called in one place, it's fine to just to change the function name in `set_error1`.

Reply:
I briefly talked to @izaid about the idea of hoisting allocations up in the ufunc loops. It's something he's thought about. He wants to be involved in the design process, but is very busy now, so hopes doing this can be put off until he has more time available to think deeply about it. He wants to come up with a general idea which will work for other array library backends as well.